src/store.rs

```diff
@@ -1253,46 +1265,84 @@ impl<F: LurkField> Store<F> {
 
     pub fn intern_scalar_ptr(
         &mut self,
-        ptr: ScalarPtr<F>,
+        scalar_ptr: ScalarPtr<F>,
         scalar_store: &ScalarStore<F>,
     ) -> Option<Ptr<F>> {
-        let tag: ExprTag = ptr.tag();
-        let expr = scalar_store.get_expr(&ptr);
-        use ScalarExpression::*;
-        match (tag, expr) {
-            (ExprTag::Nil, Some(Nil)) => Some(self.intern_nil()),
-            (ExprTag::Cons, Some(Cons(car, cdr))) => {
-                let car = self.intern_scalar_ptr(*car, scalar_store)?;
-                let cdr = self.intern_scalar_ptr(*cdr, scalar_store)?;
-                Some(self.intern_cons(car, cdr))
-            }
-            (ExprTag::Str, Some(Str(s))) => Some(self.intern_str(s)),
-            (ExprTag::Sym, Some(Sym(s))) => Some(self.intern_sym(s)),
-            (ExprTag::Key, Some(Sym(k))) => Some(self.intern_key(k)),
-            (ExprTag::Num, Some(Num(x))) => Some(self.intern_num(crate::Num::Scalar(*x))),
-            (ExprTag::Thunk, Some(Thunk(t))) => {
-                let value = self.intern_scalar_ptr(t.value, scalar_store)?;
-                let continuation = self.intern_scalar_cont_ptr(t.continuation, scalar_store)?;
-                Some(self.intern_thunk(crate::store::Thunk {
-                    value,
-                    continuation,
-                }))
-            }
-            (
-                ExprTag::Fun,
-                Some(Fun {
-                    arg,
-                    body,
-                    closed_env,
-                }),
-            ) => {
-                let arg = self.intern_scalar_ptr(*arg, scalar_store)?;
-                let body = self.intern_scalar_ptr(*body, scalar_store)?;
-                let env = self.intern_scalar_ptr(*closed_env, scalar_store)?;
-                Some(self.intern_fun(arg, body, env))
+        let tag = scalar_ptr.tag();
+        let expr = scalar_store.get_expr(&scalar_ptr);
+        if let Some(ptr) = self.fetch_scalar(&scalar_ptr) {
+            return Some(ptr);
```

Comment:
This is the final statement of the function so you can avoid `return` (see clippy).